contracts/utils/Signer.sol

```diff
@@ -113,10 +111,32 @@ contract Signer {
 
         // generate signature and add it to approvedSignatures mapping
         signature = bytes.concat(paddedAddress, bytes32(uint256(65)), bytes1(0), bytes32(uint256(messageHash.length)), messageHash);
-        approvedSignatures[keccak256(messageHash)] = 1;
+        approvedSignatures[hash] = keccak256(messageHash);
         emit SignatureCreated(signature, hash);
     }
 
+    /**
+     * @dev                Validate signature using EIP1271
+     * @param _data        Encoded transaction hash supplied to verify signature.
+     * @param _signature   Signature that needs to be verified.
+     */
+    function isValidSignature(bytes memory _data, bytes memory _signature) public virtual override view returns(bytes4) {
+        if (_data.length==32) {
+            bytes32 hash;
+            assembly {
+                hash := mload(add(_data, 32))
+            }
+            if (approvedSignatures[hash] == keccak256(_signature)) {
+                return EIP1271_MAGIC_VALUE;
+            }
+        } else {
+            if (approvedSignatures[keccak256(_data)] == keccak256(_signature)) {
```

Comment:
both checks are different, the first check checks for `approvedSignatures[hash]`. The second check checks for `approvedSignatures[keccak256(_data)]`. It's because the way, Gnosis API and Gnosis Smart Contract sends the `_data`.